WIP: make alglist return the list of algorithms

mbernasocchi opened the pull request:
This would allow checking for the presence of algorithm
```
l = processing.alglist()
if 'myNeededAlg' in l:
  do_things()
```

mbernasocchi committed:
make alglist return the list of algorithms

This would allow checking for the presence of algorithm
```
l = processing.alglist()
if 'myNeededAlg' in l:
  do_things()
```

m-kuhn commented:
I think there should be two different methods:

 * one that returns a list of algorithm names
 * one that prints a '\n' delimited string with all algs (I don't know for what, but that's what the method currently does, where is it used now?)

alexbruy commented:
@m-kuhn this method (one that prints '\n' delimited list of algorithms) is a helper method and mainly used for looking for available algs. AFAIK is not used in the Processing code.

m-kuhn commented:
@alexbruy is it expected to be invoked manually (from the console)?

What about renaming it to `dumpAlglist()` or `printAlglist()` and have `alglist()` return a list?

alexbruy commented:
@m-kuhn correct, it is usually invoked manually by user. Renaming sounds good for me, this will help to avoid possible confusion.

@volaya any objections?

volaya commented:
Sounds good to me. We can have alglist to return a list object, and then printalglist to print in the console. A user that is used to the previous method wont get an error, but the list printed instead (not so nice, but it's something...and he can then find about the new method in the docs)

Before merging this (with those changes proposed), let's make sure that we have a corresponding PR in the docs repo. Otherwise, we will most likely forget to document it...

m-kuhn commented:
@mbernasocchi feel like doing a little python today?

mbernasocchi commented:
ahhaha sorry I'm in JS mood :)

m-kuhn commented:
@mbernasocchi just delete line 52 and we have a shorter pull request queue? ;)

mbernasocchi commented:
doing a better job :)

mbernasocchi committed:
Update general.py

added possibility to get alglist, algoptions and alghelp programmatically vie the get_alg* methods
the original naming i not changed so we won't break anything

m-kuhn commented:
Thanks.

In QGIS there is (almost) no `get`-prefix. Can we keep this semantic and instead of tagging the getters with `get` tag the printers with `print`?

mbernasocchi commented:
@m-kuhn have a look if you like it and if I broke anything

mbernasocchi commented:
was a quick fix so you'r happy, need to go back to JS

m-kuhn commented on python/plugins/processing/tools/general.py:
```diff
@@ -42,17 +42,36 @@
                 s += alg.name.ljust(50, '-') + '--->' + alg.commandLineName() \
                     + '\n'
-    print(s)
+    return s
```
Does this return a string now?
Wouldn't it be more appropriate to return a `list`?

mbernasocchi committed:
made get_alglist return a list

mbernasocchi commented:
@m-kuhn renaming the original maethods to print is not an option for me right now due to time constrains, there are 40 occurences of alglist in processing and I'd need to check what these calls are. plus I'd have to document the change. I can for sure have a look at this eventually, just definetly not today

m-kuhn commented on python/plugins/processing/tools/general.py:
```diff
@@ -42,17 +42,36 @@
-                    + '\n'
-    print(s)
+                t = alg.name.ljust(50, '-') + '--->' + alg.commandLineName()
+                l.append(t)
```
Could this return a dict?

    { alg.name: alg.commandLineName() for alg in sortedList }

And the formatting would happen in the print method?

The current separation of data and representation confuses me a bit.

mbernasocchi replied:
which python version do we need to support? dict are not ordered, could use OrderedDict from 2.7

m-kuhn commented:
> there are 40 occurences of alglist in processing

fair enough, I thought there are none based on @alexbruy comments above

m-kuhn commented:
I'll leave a few comments on what I think would be good to check as well when you get to it :)

m-kuhn commented on python/plugins/processing/tools/general.py:
```diff
@@ -42,17 +42,36 @@
+    else:
+        print('Algorithm not found')
+    
+def get_alglist(text=None):
```
would be great to change `text` to something like `algname`

mbernasocchi commented:
> fair enough, I thought there are none based on @alexbruy comments above
not sure, maybe there is another method called alglist.

alexbruy commented:
@mbernasocchi if you don't mind, I will try to find time to work on this next week

mbernasocchi committed:
renamed test to alg name

mbernasocchi commented:
> if you don't mind
not at all :)

nyalldawson commented:
Superseded by #4293
